088-rebalance-progress-status.md

```diff
@@ -0,0 +1,392 @@
+
+### Field: `estimatedTimeToCompletionInMinutes`
+
+The estimated time it will take in minutes for a rebalance to complete.
+
+The formulas used to calculate the field value differ for each applicable `KafkaRebalance` state:
+
+#### State: `ProposalReady`
+
+Since the rebalance has not started at this point, we do not have the data needed to easily provide an accurate value for this field.
+Therefore, we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.
+
+#### State: `Rebalancing`
+
+The estimated time it will take for the ongoing rebalance to complete based on the average rate of data transfer.
+
+$$
+r = \frac{D_{moved}}{T_{\text{current}} - T_{\text{start}}}
+$$
+
+$$
+ETC = \frac{D_{total} - D_{moved}}{r \cdot 60000}
+$$
+
+**Notes:**
+- $r$: The average rate of data transfer in megabytes per second, calculated as the ratio of finished data movement to the elapsed time.
+- $D_{moved}$: The number of megabytes already moved by the rebalance, provided by `finishedDataMovement` field from the [/kafkacruisecontrol/state?substates=executor](#field-executorstate.json) REST API endpoint.
+- $T_{\text{current}}$: The current time when the estimate is being calculated, expressed as the number of milliseconds since the Unix epoch.
+- $T_{\text{start}}$: The time when the rebalance task was triggered, expressed as the number of milliseconds since the Unix epoch, extracted from the `triggeredTaskReason` field from the [/kafkacruisecontrol/state?substates=executor](#field-executorstate.json) REST API endpoint.
+- $D_{total}$: The total number of megabytes planned to be moved for the rebalance, provided by `totalDataToMove` field from the [/kafkacruisecontrol/state?substates=executor](#field-executorstate.json) REST API endpoint.
+- $ETC$: The estimated time to completion in minutes based on the average rate of data transfer, the value of the `estimatedTimeToCompletionInMinutes` field.
+
+#### State: `Stopped`
+
+Once a rebalance has been stopped, it cannot be resumed. 
+Therefore, there is no `estimatedTimeToCompletionInMinutes` for a stopped rebalance so we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.
+To move from the `Stopped` state, a user must refresh the `KafkaRebalance` resource, the `rebalanceProgressConfigMap` will then be updated by the operator upon the next state change.
+
+#### State: `NotReady`
+
+Once a rebalance has been interrupted or completed with errors, it cannot be resumed.
+Therefore, there is no `estimatedTimeToCompletionInMinutes` for the rebalance so we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.
+To move from the `NotReady` state, a user must refresh the `KafkaRebalance` resource, the `rebalanceProgressConfigMap` will then be updated by the operator upon the next state change.
+
+#### State: `Ready`
+
+The rebalance is complete so we hardcode the value to `0`
```

Comment:
What's the value of setting `estimatedTimeToCompletionInMinutes: 0` once you have `completedByteMovementPercentage: 100`? Can't we simply remove this field like in the other cases? It is also semantically incorrect, as this is not an estimation.

Reply:
> What's the value of setting `estimatedTimeToCompletionInMinutes`: 0 once you have `completedByteMovementPercentage`: 100

For someone tracking the `estimatedTimeToCompletionInMinutes` field alone, this is still valuable to know when a rebalance is complete.

> Can't we simply remove this field like in the other cases? 

Given we remove this field in other `KafkaRebalance` states where a rebalance cannot complete, this information is especially useful to emphasize that a rebalance is complete.

Reply:
Fair enough, thanks.